src/lib/gadgets/basic.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Basic gadgets that only use generic gates
+ */
+import type { Field, VarField } from '../field.js';
+import { existsOne, toVar } from './common.js';
+import { Gates } from '../gates.js';
+import { TupleN } from '../util/types.js';
+
+export { assertOneOf };
+
+// TODO: create constant versions of these and expose on Gadgets
+
+/**
+ * Assert that a value equals one of a finite list of constants:
+ * `(x - c1)*(x - c2)*...*(x - cn) === 0`
+ *
+ * TODO: what prevents us from getting the same efficiency with snarky DSL code?
+ */
+function assertOneOf(x: Field, allowed: [bigint, bigint, ...bigint[]]) {
+  let xv = toVar(x);
+  let [c1, c2, ...c] = allowed;
+  let n = c.length;
+  if (n === 0) {
+    // (x - c1)*(x - c2) === 0
+    assertBilinear(xv, xv, [1n, -(c1 + c2), 0n, c1 * c2]);
+    return;
+  }
+  // z = (x - c1)*(x - c2)
+  let z = bilinear(xv, xv, [1n, -(c1 + c2), 0n, c1 * c2]);
```

Comment:
Is there any reason why you used `-(c1+c2)` instead of `-c1` and `-c2` where it is a `0n` now? Meaning, efficiency-wise. Just curious, also wonder if it changes anyhow the way the term `-(c1+c2)` is encoded into the circuit.

Reply:
No reason! I could've used `-c1` and `-c2` separately as well. Yeah if you look at the function below, those constants directly become generic gate coefficients. Don't think it makes a difference efficiency-wise (but if anything, more 0s is probably better)